[JUnit] `JUnit4TestFinder` large project performance

Original discussion: https://bugs.eclipse.org/bugs/show_bug.cgi?id=540712

I have been investigating _Run As > JUnit Test_ start times for a while now.  I think there are some opportunities in `JUnit4TestFinder` to make some improvements.  

1. After `newTypeHierarchy`, all of the types from a source folder, package or project are added to the search scope.  Later, abstract or inaccessible types are excluded from the result.  Q: Is it necessary to search abstract types or anonymous types for test methods?  Is it possible in JUnit 4 to find any test cases this way?  [RxJava](https://github.com/ReactiveX/RxJava) has about 2000 test classes and over 7000 anonymous inner classes.  Minimally, I am proposing that `createJavaSearchScope` be passed only non-anonymous types.  This can be checked without fetching element info using `aClass.getElementName().isEmpty()` -- anonymous inner classes return an empty name.  The abstract test requires creating element info, so I can understand why you might wait until after.

1. After finding all test types, the `candidates` set is filtered by accessibility, abstractness and containment.  At the very least, the abstract check should be the least expensive and can be moved up front:
`if (!Flags.isAbstract(curr.getFlags()) && CoreTestSearchEngine.isAccessibleClass(curr) && region.contains(curr))`

1. Similarly in `JUnit5TestFinder` there are some expensive things like parsing within `internalIsTest`, so I propose changing the order of checks within `findTestsInContainer`:
`if (region.contains(type) && internalIsTest(type, pm))`

1. Lastly, within `JUnit5TestFinder#internalIsTest` is there a way to determine "is test" using only the `IType` passed in?  Is the `TypeBinding` the only way to make the determination?


https://github.com/eclipse-jdt/eclipse.jdt.ui/blob/master/org.eclipse.jdt.junit.core/src/org/eclipse/jdt/internal/junit/launcher/JUnit4TestFinder.java#L132
https://github.com/eclipse-jdt/eclipse.jdt.ui/blob/master/org.eclipse.jdt.junit.core/src/org/eclipse/jdt/internal/junit/launcher/JUnit5TestFinder.java#L221

After downloading the branch and working out Java 17 and Maven 3.5.4 minimums, I get this error:

[ERROR] Failed to execute goal org.eclipse.tycho:tycho-packaging-plugin:4.0.0-SNAPSHOT:build-qualifier (default-build-qualifier) on project org.eclipse.ltk.core.refactoring: No git repository found searching upwards from C:\Users\xxx\yyy\eclipse.jdt.ui-1-0b847091f80912e01032597ede8ac087a41ef00b\org.eclipse.ltk.core.refactoring -> [Help 1]

--- a/org.eclipse.jdt.junit.core/src/org/eclipse/jdt/internal/junit/launcher/JUnit4TestFinder.java
+++ b/org.eclipse.jdt.junit.core/src/org/eclipse/jdt/internal/junit/launcher/JUnit4TestFinder.java
@@ -1,5 +1,5 @@
 /*******************************************************************************
- * Copyright (c) 2006, 2021 IBM Corporation and others.
+ * Copyright (c) 2006, 2022 IBM Corporation and others.
  *
  * This program and the accompanying materials
  * are made available under the terms of the Eclipse Public License 2.0
@@ -14,8 +14,10 @@
  *******************************************************************************/
 package org.eclipse.jdt.internal.junit.launcher;
 
+import java.util.ArrayList;
 import java.util.Collection;
 import java.util.HashSet;
+import java.util.List;
 import java.util.Set;
 
 import org.eclipse.core.runtime.CoreException;
@@ -132,11 +134,20 @@ public class JUnit4TestFinder implements ITestFinder {
 			ITypeHierarchy hierarchy= JavaCore.newTypeHierarchy(region, null, new SubProgressMonitor(pm, 1));
 			IType[] allClasses= hierarchy.getAllClasses();
 
+			// filter out anonymous classes which have no name
+			List<IType> nonAnonymousClasses= new ArrayList<>();
+			for (IType t : allClasses) {
+				if (!t.getElementName().isEmpty()) {
+					nonAnonymousClasses.add(t);
+				}
+			}
+			IType[] filteredClasses= nonAnonymousClasses.toArray(new IType[0]);
+
 			// search for all types with references to RunWith and Test and all subclasses
-			HashSet<IType> candidates= new HashSet<>(allClasses.length);
+			HashSet<IType> candidates= new HashSet<>(filteredClasses.length);
 			SearchRequestor requestor= new AnnotationSearchRequestor(hierarchy, candidates);
 
-			IJavaSearchScope scope= SearchEngine.createJavaSearchScope(allClasses, IJavaSearchScope.SOURCES);
+			IJavaSearchScope scope= SearchEngine.createJavaSearchScope(filteredClasses, IJavaSearchScope.SOURCES);
 			int matchRule= SearchPattern.R_EXACT_MATCH | SearchPattern.R_CASE_SENSITIVE;
 			SearchPattern runWithPattern= SearchPattern.createPattern(Annotation.RUN_WITH.getName(), IJavaSearchConstants.ANNOTATION_TYPE, IJavaSearchConstants.ANNOTATION_TYPE_REFERENCE, matchRule);
 			SearchPattern testPattern= SearchPattern.createPattern(Annotation.TEST.getName(), IJavaSearchConstants.ANNOTATION_TYPE, IJavaSearchConstants.ANNOTATION_TYPE_REFERENCE, matchRule);
@@ -147,7 +158,7 @@ public class JUnit4TestFinder implements ITestFinder {
 
 			// find all classes in the region
 			for (IType curr : candidates) {
-				if (CoreTestSearchEngine.isAccessibleClass(curr) && !Flags.isAbstract(curr.getFlags()) && region.contains(curr)) {
+				if (!Flags.isAbstract(curr.getFlags()) && CoreTestSearchEngine.isAccessibleClass(curr) && region.contains(curr)) {
 					result.add(curr);
 				}
 			}

--- a/org.eclipse.jdt.junit.core/src/org/eclipse/jdt/internal/junit/launcher/JUnit5TestFinder.java
+++ b/org.eclipse.jdt.junit.core/src/org/eclipse/jdt/internal/junit/launcher/JUnit5TestFinder.java
@@ -1,5 +1,5 @@
 /*******************************************************************************
- * Copyright (c) 2016, 2021 IBM Corporation and others.
+ * Copyright (c) 2016, 2022 IBM Corporation and others.
  *
  * This program and the accompanying materials
  * are made available under the terms of the Eclipse Public License 2.0
@@ -218,7 +218,7 @@ public class JUnit5TestFinder implements ITestFinder {
 
 		// search for all types with references to RunWith and Test and all subclasses
 		for (IType type : allClasses) {
-			if (internalIsTest(type, pm) && region.contains(type)) {
+			if (region.contains(type) && internalIsTest(type, pm)) {
 				addTypeAndSubtypes(type, result, hierarchy);
 			}
 		}

